crates/circuit/src/operations.rs

```diff
@@ -323,71 +329,415 @@ impl Operation for OperationRef<'_> {
         }
     }
     #[inline]
-    fn control_flow(&self) -> bool {
+    fn directive(&self) -> bool {
         match self {
-            Self::StandardGate(standard) => standard.control_flow(),
-            Self::StandardInstruction(instruction) => instruction.control_flow(),
-            Self::Gate(gate) => gate.control_flow(),
-            Self::Instruction(instruction) => instruction.control_flow(),
-            Self::Operation(operation) => operation.control_flow(),
-            Self::Unitary(unitary) => unitary.control_flow(),
+            Self::ControlFlow(op) => op.directive(),
+            Self::StandardGate(standard) => standard.directive(),
+            Self::StandardInstruction(instruction) => instruction.directive(),
+            Self::Gate(gate) => gate.directive(),
+            Self::Instruction(instruction) => instruction.directive(),
+            Self::Operation(operation) => operation.directive(),
+            Self::Unitary(unitary) => unitary.directive(),
         }
     }
-    #[inline]
-    fn blocks(&self) -> Vec<CircuitData> {
+}
+
+/// Used to tag control flow instructions via the `_control_flow_type` class
+/// attribute in the corresponding Python class.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[pyclass(module = "qiskit._accelerate.circuit", eq, eq_int)]
+#[repr(u8)]
+pub(crate) enum ControlFlowType {
+    Box = 0,
+    BreakLoop = 1,
+    ContinueLoop = 2,
+    ForLoop = 3,
+    IfElse = 4,
+    SwitchCase = 5,
+    WhileLoop = 6,
+}
+
+#[derive(Clone, Debug, IntoPyObject, PartialEq)]
+pub enum BoxDuration {
+    Duration(Duration),
+    Expr(expr::Expr),
+}
+
+#[derive(Clone, Debug)]
+#[repr(align(8))]
+pub enum ControlFlow {
+    Box {
+        duration: Option<BoxDuration>,
+        annotations: Vec<Py<PyAny>>,
```

Comment:
For a follow-up: we'll most likely want to conditionally compile out `annotations` when in stand-alone C mode to avoid the `Py` symbols.